lib/read.js

```diff
@@ -182,68 +174,81 @@ var Read = Juttle.proc.source.extend({
-            this.handleDebug();
-            return;
-        }
+        this.queueSize = limit;
 
-        function executeQuery() {
-            if (self.raw) {
-                return self.baseQuery
-                .then(function(res) {
-                    return self.handleRawResponse(res);
-                })
-                .then(function(points) {
-                    return self.formatAndSend(points);
-                });
+        if (this.raw) {
+            queryExecutionPromise = this.baseQuery
+            .then((res) => {
+                return this.handleRawResponse(res);
+            })
+            .then((points) => {
+                return this.formatAndSend(points);
+            });
+        } else {
+            var query = this.baseQuery.clone();
+            if (this.timeField) {
+                if (from) {
+                    query = query.where(this.timeField, '>=', new Date(from.valueOf()));
+                    this.logger.debug('start time', new Date(from.valueOf()).toISOString());
+                }
+                if (to) {
+                    query = query.where(this.timeField, '<', new Date(to.valueOf()));
+                    this.logger.debug('end time', new Date(to.valueOf()).toISOString());
+                }
             }
 
-            if (self.batchBuckets) {
-                return self.paginateBuckets();
+            if (this.debugOption) {
+                queryExecutionPromise = Promise.try(() => { return this.handleDebug(query); });
+            } else if (this.optimization_info.reduce_every) {
+                this.getBatchBuckets(this.optimization_info, from, to);
+                queryExecutionPromise = this.paginateBuckets(query);
+            } else {
+                this.offsetCount = 0;
+                queryExecutionPromise = this.paginate(query);
             }
-
-            self.offsetCount = 0;
-            return self.paginate(self.baseQuery.clone());
         }
 
-        executeQuery()
-        .catch(function(err) {
+        return queryExecutionPromise
+        .catch(err => {
             if (/Pool (is|was) destroyed/.test(err.message)) {
-                var connectionInfo = self.baseQuery.client.connectionSettings;
-
-                self.trigger('error', self.runtime_error('RT-INTERNAL-ERROR', {
-                    error: 'could not connect to database: ' + JSON.stringify(connectionInfo)
-                }));
+                var connectionInfo = this.baseQuery.client.connectionSettings;
+                this.trigger('error', new Error('could not connect to database: ' + JSON.stringify(connectionInfo)));
             } else {
-                self.trigger('error', self.runtime_error('RT-INTERNAL-ERROR', { error: err.message.toString() }));
+                this.trigger('error', err);
             }
-        }).finally(function() {
-            self.eof();
+        })
+        .then(() => {
+            var points = this.pointsBuffer;
+            this.pointsBuffer = [];
```

Comment:
If pointsBuffer is an instance variable what happens if there are multiple calls to read at once from different points in the flowgraph? Won't they stomp on each other?